e2e framework: test labels

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This is an implementation of the idea outlined in
https://github.com/kubernetes/enhancements/pull/3042/files#r988729747.
    
It becomes possible to select tests in different ways:
    
        ginkgo -v -focus="should respect internalTrafficPolicy.*\[FeatureGate:ServiceInternalTrafficPolicy\]"
    
        ginkgo -v -label-filter="FeatureGate:ServiceInternalTrafficPolicy"
    
        ginkgo -v -label-filter="Beta"
    
When the test runs, ginkgo shows it as:
    
        [It] should respect internalTrafficPolicy=Local Pod to Pod [FeatureGate:ServiceInternalTrafficPolicy] [Beta] [FeatureGate:ServiceInternalTrafficPolicy, Beta]
    
The test name and the labels at the end are in different colors. Embedding the
annotations inside the text is redundant and only done because users of the e2e
suite might expect it. Also, our tooling that consumes test results currently
doesn't know about ginkgo labels.

#### Notes for reviewers

Please look at this commit-by-commit.

For an overall impression on how to use the new APIs, see test/e2e/framework/internal/unittests/bugs/bugs.go.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
e2e: add wrapper functions to annotate tests

These wrapper functions set labels in addition to injecting the annotation into
the test text. It then becomes possible to select tests in different ways:

    ginkgo -v --focus="should respect internalTrafficPolicy.*\[FeatureGate:ServiceInternalTrafficPolicy\]"

    ginkgo -v --label-filter="FeatureGate:ServiceInternalTrafficPolicy"

    ginkgo -v --label-filter="Beta"

When a test runs, ginkgo shows it as:

    [It] should respect internalTrafficPolicy=Local Pod to Pod [FeatureGate:ServiceInternalTrafficPolicy] [Beta] [FeatureGate:ServiceInternalTrafficPolicy, Beta]

The test name and the labels at the end are in different colors. Embedding the
annotations inside the text is redundant and only done because users of the e2e
suite might expect it. Also, our tooling that consumes test results currently
doesn't know about ginkgo labels.

Environments, features and node features as described by
https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/3041-node-conformance-and-features
are also supported.

The framework and thus (at the moment) test/e2e do not have any pre-defined
environments and features. Adding those and modifying tests will follow in
a separate commit.

pohly commented on test/e2e/network/service.go:
```diff
@@ -2517,7 +2518,7 @@ var _ = common.SIGDescribe("Services", func() {
 	})
 
-	ginkgo.It("should respect internalTrafficPolicy=Local Pod to Pod [Feature:ServiceInternalTrafficPolicy]", func() {
+	framework.It("should respect internalTrafficPolicy=Local Pod to Pod", framework.WithFeatureGate(features.ServiceInternalTrafficPolicy), func() {
```
We could also accept "ServiceInternalTrafficPolicy" as string here. But then we loose the compile-time verification that the string is actually a valid feature gate. It would still get checked at runtime whenever the suite gets invokved.

The advantages would be:
- no need for the extra package import
- shorter
- no need to change anything here when a feature graduates and (later) gets removed from k8s.io/kubernetes/pkg/features (but see the KEP discussion how the test should get annotated for a stable feature)

aojea replied:
I prefer to maintain the link to the code, to avoid drifting

pohly replied:
That's also my preference. Using the pre-defined string constant also has the advantage that tools like gopls can reliably find tests that reference it.

pohly replied:
I extended the PR with sets of valid features and environments. The framework itself doesn't register any of those, this will have to be added elsewhere (test/e2e/environments and test/e2e/features?).

pohly replied:
I added pre-defined features and node features in their own shared packages (`test/e2e/feature`, `test/e2e/nodefeature`). The names were chosen so that using these constants reads well. Tests are not allowed to add their own features or node features (registry gets frozen after those pre-defined ones).

There is nothing for environments yet because that's currently unused.

We have to allow arbitrary labels - there are simply too many of those and they might be useful. Those arbitrary labels don't get registered. That leads to a loophole: `framework.WithLabel("Feature:SomeCustomFeature"` is currently valid, but should be `feature.SomeCustomFeature` with a corresponding entry in the shared package. I think we can check those custom labels and complain when one of the other functions should be used instead.

pohly replied:
We also need a way to verify that tests are registered properly before we can merge this. There are some existing verify scripts that need to be updated. My plan is to do the verification as much as possible inside the E2E suite itself:
- Validate labels when functions are called.
- Use https://github.com/onsi/ginkgo/issues/1225 once it is available to add `--verify-tests` which then can be called by a verify script.

pohly replied:
I've completely revamped how issues discovered during test registration are handled. Each E2E suite now records such problems, then at the end does a sanity check that there were no issues. This works for `-dry-run` and `-focus=some tests`, so it'll be impossible to miss.

A verify script also checks this as part of `pull-kubernetes-verify`.

aojea commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -31,3 +37,114 @@ func AnnotatedLocation(annotation string) types.CodeLocation {
+
+// WithFeatureGate adds a certain feature gate as label to a [framework.It].
+// The feature gate must be listed in
+// [k8s.io/apiserver/pkg/util/feature.DefaultMutableFeatureGate].
```
this will solve the problem of feature promotion, that is common to left e2e tests behind and not update them ... I think is something we should do

aojea commented:
/assign @spiffxp @aojea

aojea commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -31,3 +37,114 @@ func AnnotatedLocation(annotation string) types.CodeLocation {
+	// the empty string.
+	var level string
+	if spec.PreRelease == "" {
+		level = "Stable"
```
my vote is about not tagging stables at all, as they must be part of the core system I don't see the value in the feature gate tags

pohly replied:
I agree. I added it only because https://github.com/kubernetes/enhancements/pull/3042 has it. The actual usage is not clear to me yet.

SergeyKanzhelev replied:
this is for those rare cases when feature gate is GA, but not locked to default.

pohly replied:
So a GA feature that can still be disabled? I can see why continuing to list such a feature as test tag may be useful.

But how does marking such a feature as "Stable" help? In other words, how would the "Stable" tag be used? If it's not used, we don't need it...

SergeyKanzhelev replied:
I'm not attached to the tag name. Just thought we need something so we can find jobs that set the `FeatureGate`, but haven't set it's stage

pohly replied:
My proposal in this PR is to derive the stage from the feature, i.e. it never would be set explicitly.

pohly replied:
But we still need to clarify what's supposed to happen when a feature graduates to GA and then later gets removed from the set of defined features.

spiffxp replied:
>my vote is about not tagging stables at all

Agree, I feel like this could imply `[Stable]` should be added to everything that is currently not `[Alpha]`, `[Beta]` or `[Deprecated]` and I don't see the value in it.  Also `featuregate.GA` is already `""` so we're keeping parity in some sense.

> this is for those rare cases when feature gate is GA, but not locked to default.

Some examples to motivate discussion: https://github.com/kubernetes/kubernetes/blob/5113b705d2b2c96f04f9eca5ad427ba3b441ae23/pkg/features/kube_features.go#L872-L879

> Just thought we need something so we can find jobs that set the FeatureGate, but haven't set it's stage

regex or label equivalent to `+FeatureGate:.* -Alpha -Beta -Deprecated` but again I'm not sure why we would want or need this. 

If we set the convention that everything uses `WithFeatureGate` which auto-derives alpha/beta/deprecated I'm not sure this matters so much?

>we still need to clarify what's supposed to happen when a feature graduates to GA and then later gets removed from the set of defined features

Same thing, no stable/GA label? I think you're talking Feature, not FeatureGate here, and I'm not entirely sure why the difference matters.

pohly replied:
I was talking FeatureGate and how it transitions (or doesn't?) to Feature.

Let's look at a fictional test that depends on FeatureGate foo. While that is in alpha/beta, we annotate the test with `[FeatureGate: foo] [alpha/beta]`, with '[alpha/beta]' derived from the definition in `kube_features.go`.

Now it graduates to GA and remains in `kube_features.go`, either because it isn't locked to its default yet or because all features are kept around a bit longer to avoid breaking code that uses `--feature-gates foo=true`. During that time we annotate the test with `[FeatureGate: foo]` and nothing else.

The last step is removal from `kube_features.go`. Now we remove `[FeatureGate: foo]`. We have to do something, because it won't compile anymore.

Or do we turn it into `[Feature: foo]`? In all cases or only where it matters? It matters if the feature, even though GA, is not available everywhere.

For features that are available everywhere, adding `[Feature: foo]` has the advantage that it becomes easy to run all tests related to a certain feature. The downside is that many jobs currently exclude tests marked with `[Feature: <something>]` and would therefore exclude the feature. I'm undecided.

aojea replied:
I think that we should not create this relation Feature <---> FeatureGate, those are independent things that some times be related, a FeatureGate doesn't have to be related to a Feature necessarily.

pohly replied:
So once a feature gate gets removed from `kube_features.go`, we also remove the test annotation. Works for me.

pohly commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -31,3 +37,114 @@ func AnnotatedLocation(annotation string) types.CodeLocation {
+// TODO: Describe and Context wrappers.
+
+func WithFeature(name string) withFeature {
+	// TODO: verify that the feature is known.
```
@aojea: do you agree that `WithFeature` and `WithEnvironment` should only accept "known" values?

We would have to create a registry similar to `FeatureGate` and add all values that are currently used in-tree to it by default.

pohly replied:
@SergeyKanzhelev how do you want to continue with https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/3041-node-conformance-and-features?

pohly commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -31,3 +37,114 @@ func AnnotatedLocation(annotation string) types.CodeLocation {
 }
+
+// It is a wrapper around ginkgo.It which supports [WithFeatureGate],
+// [WithFeature], [WithEnvironment], [WithConformance], [WithNodeConformance]
```
@aojea: any preference from your side here?

dims commented:
If you still need this PR then please rebase, if not, please close the PR

pohly commented:
/reopen

I intend to add these helpers and perhaps use them for some tests, but without breaking the current naming of tests.

pohly committed:
e2e framework: list specs and labels

-list-tests is a more concise alternative for `ginkgo --dry-run` with one line
per test. In contrast to `--dry-run`, it really lists all tests. `--dry-run`
without additional parameters uses the default skip expression from the E2E
context, which filters out flaky and feature-gated tests. The output includes
the source code location where each test is defined. It is sorted by test
name (not source code location) because that order is independent of
reorganizing the source code and ordering by location can be achieved with
"sort".

-list-labels has no corresponding feature in Ginkgo.

One possible usage is to figure out what values might make sense for
-focus/skip/label-filter.

Unit tests will follow in a future commit.

pohly committed:
e2e: define features and node features

The list is based on the -list-tests output.

pohly committed:
e2e: enhance SIGDescribe

framework.SIGDescribe is better because:
- Ginkgo uses the source code location of the test, not of the wrapper,
  when reporting progress.
- Additional annotations can be passed.

To make this a drop-in replacement, framework.SIGDescribe generates a function
that can be used instead of the former SIGDescribe functions.

windows.SIGDescribe contained some additional code to ensure that tests are
skipped when not running with a suitable node OS. This gets moved into a
separate wrapper generator, to allow using framework.SIGDescribe as intended.
To ensure that all callers were modified, the windows.sigDescribe isn't
exported anymore (wasn't necessary in the first place!).

alexzielenski commented:
/triage accepted

pohly commented on test/e2e/cloud/gcp/addon_update.go:
```diff
@@ -241,7 +241,7 @@ var _ = SIGDescribe("Addon update", func() {
 
 	// WARNING: the test is not parallel-friendly!
-	ginkgo.It("should propagate add-on file changes [Slow]", func(ctx context.Context) {
+	framework.It("should propagate add-on file changes", framework.WithSlow(), func(ctx context.Context) {
```
All of these `framework.` package selectors are bit lengthy.

What do you think about exposing these functions also as methods of the `Framework` struct, even though they are not using that struct?

The we could write here:
```
f.It("should propagate add-on file changes", f.WithSlow(), func(ctx context.Context) {
```

aojea replied:
you can do it just changing the import `f "k8s.io/kubernetes/test/e2e/framework"` , can't you?

I don't like the idea of creating this dependency just to shorten text , it seems like will add more overhead to maintain it

pohly replied:
`f "k8s.io/kubernetes/test/e2e/framework"` would imply replacing all usages of `framework` with `f` and require changing the default name `f` in  `f := f.NewDefaultFramework()` because the variable `f` would shadow the package. That's a massive change.

pohly replied:
> more overhead to maintain it

We have to weigh that against the overhead of more typing and longer lines in all calls of them.

My expectation is that these functions should only need to be implemented once with little future changes, but they will have to be typed a lot.

pohly replied:
We also cannot do:
```
import (
    "k8s.io/kubernetes/test/e2e/framework"
    f2 "k8s.io/kubernetes/test/e2e/framework"
)
```
because there is a linter check for double imports.

We could move the new functions into `k8s.io/kubernetes/test/e2e/framework/ginkgo` and import it as `fg`. But that means that all test files will need to import both, which feels worse.

pohly replied:
This is easier to discuss based on the actual code.

I added the `Framework` methods and one separate commit at the end which uses them. We can still back that out if this isn't desirable.

pohly commented on test/e2e/apimachinery/validatingadmissionpolicy.go:
```diff
@@ -31,11 +31,12 @@ import (
 )
 
-var _ = SIGDescribe("ValidatingAdmissionPolicy [Privileged:ClusterAdmin][Alpha][Feature:ValidatingAdmissionPolicy]", func() {
+var _ = SIGDescribe("ValidatingAdmissionPolicy [Privileged:ClusterAdmin][Alpha]", feature.ValidatingAdmissionPolicy, func() {
```
Do we want to allow tests to register tags like `[Privileged:ClusterAdmin]` as plain text? Using `framework.WithLabel` would be better.

If we require it, then we can enforce that the plain text strings don't contain any `[]` and thus catch when developers add those back by mistake.

pohly replied:
I discussed this with @aojea on chat and the conclusion was to leave this as-is for now to minimize the amount of changes that we do in a single PR.

I can probably start doing some lighter validation for `[]` inside the text and catch the ones that shouldn't be added that way anymore, like `[Alpha]`.

aojea replied:
yeah, this changes have always a big blast radius and the changes take a long time in be detected so we should be conservative on the rollout of this feature

pohly commented on test/e2e/framework/test_context.go:
```diff
@@ -356,6 +359,8 @@ func RegisterCommonFlags(flags *flag.FlagSet) {
 
 	flags.BoolVar(&TestContext.ListImages, "list-images", false, "If true, will show list of images used for running tests.")
+	flags.BoolVar(&TestContext.listLabels, "list-labels", false, "If true, will show the list of labels that can be used to select tests via -ginkgo.label-filter.")
+	flags.BoolVar(&TestContext.listTests, "list-tests", false, "If true, will show the full names of all tests (aka specs) that can be used to select test via -ginkgo.focus/skip.")
```
`-list-tests` and `-list-labels` work as implemented in this commit, but using https://github.com/onsi/ginkgo/issues/1225 would make the output cleaner.

Should I take it out for now or do we want to merge it as-is and improve it later?

I was using `-list-tests` to check for changes in the test texts caused by this PR. There are some because in some cases, I unified the placement of tags and insertion of spaces.

pohly replied:
Example:
```console
$ go test -v ./test/e2e -args -list-tests
=== RUN   TestE2E
  I0705 10:04:42.204887   72879 e2e.go:117] Starting e2e run "7285667e-32ed-4bca-90cc-e7dfcc2a8fee" on Ginkgo node 1
[1688544282] Kubernetes e2e suite - 4644/7346 specs •••••S•S•S•S••S••••SS••S••S•S•SSS••S••S•••S•••••••S•S•S•SS•SSSS•S•S•S••S•SS••S•S••S•••SSSS•••••••••••S••S•••S••••S••S••S••••••S••SSS••••S••••••••S•S••S•S••••S••••SSSSSS•••••SS•S•••••••••SS••••••SS•S•S••••SS•S••••••••SS••S••••••••S••••S••S•S•S•S••S•S•S••SSS••S•S•SS••••••S••SS•S••SS•SSSS•••••S••S••••••••••••••SSS•S••••SS•S•••••S•SS••S•••S•••••••••S•S•S•SSS•S•••SS•SSS••SS•••S•S•S••SS•••SS•••S•S•S••••S••S••S••S•SS•S••S•••SSSS•S••S•••••SSS•S•S•••••••SSS••••••••S•S•••S•••••S••••••S•S••••••••S•••SS••S•S•S•S••S•S•SS••S•••••S•SSS•S••SS•••••••SS••••••••••S••SSSSSSSS•••••SSS•S•••SSS•S•S•••SS•S•••SS•••••••••S•SS••SSSS•••••S•SSS•••S•SS••••••SS••SS••S•SS•S•••SS•••S••SSS••SS•S•S••SS•S•••S•••••••S•SS••••••••S•S•S•••S•SSS•S••S•••SS•S•SS•••••S•S•S••S••S••••••••••••••S••••••••SSS•S•S•SS•SS•SSSS••SSSS•S•S••S•••SSSSS•••••••S•SS•S•S••S••S•S••S•S••S••••S••S••S•••••SS•S•••S••SS••SS•••••S••SSSS•S•S•S•••••••S••SSS••••SSSS•••••SS•S•••S•••••••S••SS••S•SSSS••S••••••••••S•••S•••••SS•S•SS••S•••S•S•S•••S••SS•S•••••••S••S••S•••••••S•••S••S•S••SSSS•S•S•SS••••SS••SS•••••••••S•••S••S••S••••S••••S•••S•S••SS•••S•SS••SSS••SS••••••S••S•S•••••••••S•S•SS•S•••S••••S•••S•••S••S•SS•••S•••SSS•SSSSS•••S••S••S•••••S••S••••SSSS•••S••••S•••S•S•••SSS•••SS•SS•••S•••SS••••SS•S••SS••••S•••••S••••SSS•••••••••S•SS••SS•SS••••SS••SS••S•S••SSS•••S•••••SS•SSS••SS•S••••••SS•••S•SS•S••SS•••SS•S•••SS••S•••••S••S•SS•SSS•SSSS•S••SS••S•S•••SSSS••S••••S•••S•S•SSSS•••SS•••••••SS•••S•••S•SS••••S•S••S•••S••S••••••••S•••S•SSS•••••••SSSS••S•••••S••SSS••SSS••SS••SSS•S••SSS•••S•••••S•SSS••S•SS••SS•SS•S••SSS•••S•S•SS••SSS•SS•S•S•S•SS•SS•SSSS•••S•S•SS••••SS•••SS•••••••••S•S•S•••••••S••••••••••••••S•••S•••••••••••••S•S••••S••S••S•••S••SS•S•••S••SS••••SS••••SSSS•S••••S••SSS••S•S•SS•••S•••••••S•••••••••SS•••SSS•S•••••SS•S••S•••SS••SS•S•••••••••S••SSS••••S••S••S•SS•••••S••••SS•SSS•••S••S•SS•••••SSS••S••••••••SS••••••SS••••SSS•SSS••S••••SS•S•••••••S•SSSS•S•••••SSSS••SS••••••S•••S•••S•••S•••SS•••SS•••••S•SS••••••S•••S•••••••S•••SS•SSSSS•••SS•S•••••S•SS••SSS•••SSS•SS••S•••S•••••••SSSSS•S•S•S•••••••S•S••••SS•••S••S••••••S•S•••••••••S•S••••S•SS••••SSSS•••••SS••SS•SSSS•S•••S••S••S•••S•••SS••S••••••SS•SS•••S••••SS•••S••SS••S•S••S•••SS••S••S•••••SS•S•••••S••••S•S•••••SS••••••••S•••S••••S•••S••S•••S•S•SS•S•SS•S•S••SS•S••S••S••S••••••SS••SSSS••••••S••••S•••••SS•••••SS••••••S•S•••SS••S•S•S••••S•SSSS•S•S•••••••S•••••S••S•S•S•S••S•••••S••SS•SS•S••••••SS••••SSS••••••SSSSSS•SSS•S•••S••••••S••S•S•S•S••••S•••••••S••SS•••S••S•••••S••SS•••••••S•••S•SSSS•S•SS•••S••S••S•••SS••S••••S••S••••S•••S•••••••S••S•••••SSS••SS•SS••••S•S•••S•••••SS•SSS•••S••S•S•S••SSS•••SSS•SS•S••SS•SS•••••SS•SSSSSSSS•S•SS•S•S•S•S•S••SSS•••S•S••S••S••••S••SS••••••••••••SSS•S•S••S•••S•••S•SS•S•S•S•S••••••SS•••••SSS•S••••••S•••SS••S••SS••••SSS•••S••S•S••••••••••••S•SSS•S•••••S•S••••S•••••••S••SSS•SS••••••••••S••••S•SS••••••••S•••S•S•••SS••SS•S•••S•SSS••SSS•••S•••S•S•SS•S••SS•S•••••••••S•SS••SS••SS•SS•••••••S•••••S•••SS•S•SS•••SSS•••••SSS•SS••••S•••S••SSS••S••••S•••••••••••SSS•••S••S•S•••••••••SSS•SSS•S•••S•S•••SSS•S••••••••S•••••••S•SS••SSS•••••S••••••S••S••SSS••SS•S•••••S•SS•S••SS••S•S•SSS•S•S••SS•••S••••••S•S••SS•••S••••SS••••S•••S••S•••••SS••SS•••S•••••••••••••S•••SS•S•SS•S••••SS•••SS•••••S•S•••••SS••S•S•SS•••S••••SS•S•S•••S••SS•••S•••••••SSS•S••SS••••SS•SSS••S•S•SS••••SS•••S•SS•••S•SS•••••S•S•••••SS•S••S•S•S••S•SSS•••S••S••••S•S••SSS••••SS•S••SS•••S•••S•••SS•SS•S•S•S•S•SS••S•S•••••SS•SSS•S•S••S••SS•••S••••S•S•S••••••SS•••SSS•S••S••SS••••S••••S•••S•••••S•••••S•••S•SS•SSS••S••S••••••S••S•S••S••SS•••S•••S••SSSS•••S•••S•SS•••S••S•S••S•S•••••SSS•S••S••SS••S•SS•SS•••S•S•••••••SS•••S•SS•SS•SS•S•••••••S••S••S••••••S•SS••S•SSS••SSS•S•••••SS••SS••SSS•SSS•••••S••S•SSS•S••••S••SS•••S•SS•••••S•••S•••••••••SS••••S•••SS••SS•S•S•••S••S••S••SS••••••S•SSS•••SS•••••SSSS•S••S•SSSSS•••S••••S••SS••SSS•••SS••S••S••SSS•S•••••••••SS••S•••S•S•••S••S•••S•••SSS•S•S••S••S•SSSS••S•SSS•••S•••S•SS••••••S•S••S•S••••SS•S•••S•S•••••S•••S•SS•SS••••S•S••S•S•S••••S•SSS•••SS••SS•S•••S•••S••••••S•••SS•SSSS•S•SS•S••SS•SS•••••SSSSS••••••S•••S•S••••SSS•••••SS•SS•S•••••SS•SSS••SSS•••SSS•••S•SSS•••SSS••••S••S••S••••••••••SS•SSSSS••••••S••S••••SS•SS•SS•SS•••SS••SS•••S••S•SSSSS•SS•S•••S••SS•SSS•••S•••••SS••S•••S•SSS•S••SS•S••S•••••SS•SS••S•S•S•••SS••••S••••SS•S•S•S••••S•SS••SS••S••S•S•••••••••S••S••••••SSS•S••SSS•S•••S•S••••SSSSSS•••S•SSSSS•••S••••••••S••••SSSS••SSS•SSSS•S•SS•SSSS•SSS••SS•••SS••S•S••••S•••••S••••••S•••SSSSSS••SS••S••••SS•S•SSSSS••SS•••S•••S•SSS•••S••SS•S•S••SS••S••S••••••••S•SS••••SS•S•SS••••SS•S•••S••SS••••SSS•••S•••••••••S••S•S••S•S•••S•S••••••••••SS•••••••S••••S•S••S•S•••S•••••••S•••S•S•S•••••••S••S••••••••S•••••S•SSS•••S•SSS•••••SS••S•SS•S•S•••••••••••••S•S•S•••S•••••SS•SS•S•SS••S••S•S•••SSS•S••••S•S•••SS••S•••••••••SSS••••••SS••••••••••••S••S••S•••S••••••••SS•S••••SS•S••••••••SS•••••••••SS•••••SSS•••••••SSS••S••S•S••••SS•S•••SSS••SS••S•••S•SSS•••SS••S•S•SS•••SS••S•••••••S•••S••••SSS••SS•S••SSSS•SSS•S••••S•SS•••S•••••••S•S••S•••S••S•••S•••S••••SSSS•S••SSS•S•••S•••S••SS•SS••••••••••S••SS••••••••S•S••S•S•SSS••S••••••••SS•••S••S•••S••••S•S••S•S•SSS••••••••••S••S•••S••SS•S••••SS•S•S•SSSS•S•S•••••••••••••••S•••••S•SSSSS••S•••S•SS•S•SSSS••S••S••S•S•••SS••••••••••S•S•S•••S•SSSS••••SSSSS••S•SS•••S•S•S•SSS•SS•SSS•••S••••S•SS••••S•S•SS•S•••S••S•••SS•SS••••SS••S••S•SS•S•SS••••••S••SS••••S•S•S••S•S•••S•••••S•••S•SSSS••SS•SS•S•S••SS•SS••••••SS•••••••SSSS••SS••••S••S••SSS••••S••••SSS••••••S•S•S•••S••S••SS•SS••SS••••••SS••••••S•S••SS•••SS••SSS•••SSS•SS••S•••••S••••••S•••••S••••S••SS••••••S•••SSS•SS•SS•S••SSS•••S••SS•SS•••••••••••••••S••••S•SSS•SS•SS•••S••S•••S•SS•SSSSSS••S•••••S•••S••SS••••••••••S••••S••••S••S•••S••••S•••SSS•••S•••S•S••••S•••••••SSS••••••S••••••S••S••SS•S••••••••SSS•SS•••SS•••••S•S••S••••••••S•S••S•SSS•••SS•••S•S••••S•SS•SS••S•S•S••S•S•SS•S•••S•SSSS••••••S••••S••••S•SS•S••S••••S•SSS••S•S••••••••SS•S•SS•S••SS••••••S•S••••S•••S•••••SS•S••••S•SS•S•S•S•••S•S•S••SSS•••S•S•S••SS•SS•SSSS••S•••••••••••••••••••••••••SSSSS•••S•S•S•S•••••SSS•SS••S•S•S••••SS••S•SS••SSS••S••S•S••SS••SS•••••••S•S••••S••S••S•S••S•••S•••S••S•SSS•S••••••S•••S•••••SS•S••SS••S•S•S••S•SS•S••S••S•S••S•SSS•S••S••SSSSS•••S•SS••••SS•••••••••S••S•S••S••SSSS••••SSSS•••••S••••••••S••S••SS•S•S••SS•SS••SS•••S•S••S••S••S•SSS••S••••••SS••••••••SS•S•SS••SSS•••••S••SS•SS•S•S••••••S•S•S•SSS•S••SS••SS•••••SS•••SS•S•S•••S•S•SS••SSS•SSS•••••••••S•••••S•••S••••S••SS•••SS•••SS•S•S•S••S•SS•••SS••SS•••SS•S•••S••SS••S••••S•SS••••••SS•••S••SS••SSS•••S•••••••SS•••••S••••••S•••S•••••••S•••••••SS•SSSSS••••••S••SS••••SSSS••••S•••••••••••••S•S•S•S•S•••S•••••S•••S•••SSS•S•SSS••••••SS•S•S••SSSS•••SSSS•••••SS•••SS•••••S•S•••S••SS•S•••S••••S•••S•••S••S•S••S••S•S•SS•S•SS•S••••••••SS•S•S•••S•••S••••S•••SS••SS•SS••••••••SS••S•S•SS•SS••SS•SS•S•••••S•SS••••S••••••S••••S••••SS••••SSSS•••S•S•S•SS•SS••••SS••SSS•••SS•••••S••S•••••••SSS•••S••S••••••S•SS•S•SS•S•••S••••S•••SS••••••S•S••••••S••S•S••S••S•SS••••SSS•S•S•SSS•S•SSS•••••••SS•S••S•S•S•S•S•SSSSSS•••••••S•SSS••••S•••••••SS•S•S•S••SSS•••S••S••SSS•••S•S•SSSSSS•••••••S••SS•S•S••••••S•S•S••••••••••SSS•••••••S•S••SS•••••S••S••SS•S•SSSS•S••••S••••SS•SS•SS•••S••••S•••S•SS•••••••SSS••••••••••••S•S•••••S••S•S•S•SSS•••••••••S•SS••SS•••S••SS••SSS••••S••S••••SS••••S••••SSSS•••SSS••S•••••••S•S•SSS••SS•S•S•S•••••••••S•••S•SS•SS•S•S•••••••S•••••SS•SSSS•SSSS•S•••S••SSS•S••••••SS••••••SSSSS•••S

The following spec names can be used with `ginkgo run --focus/skip`:
    [sig-api-machinery] API Streaming (aka. WatchList) [Serial] [Feature:WatchList] should be requested when ENABLE_CLIENT_GO_WATCH_LIST_ALPHA is set
    [sig-api-machinery] API priority and fairness should ensure that requests can be classified by adding FlowSchema and PriorityLevelConfiguration
    [sig-api-machinery] API priority and fairness should ensure that requests can't be drowned out (fairness)
    [sig-api-machinery] API priority and fairness should ensure that requests can't be drowned out (priority)
    [sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin] listing mutating webhooks should work [Conformance]
    [sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin] listing validating webhooks should work [Conformance]
    [sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin] patching/updating a mutating webhook should work [Conformance]
    [sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin] patching/updating a validating webhook should work [Conformance]
...
    [sig-windows] Services should be able to create a functioning NodePort service for Windows
    [sig-windows] Windows volume mounts [Feature:Windows] check volume mount permissions container should have readOnly permissions on emptyDir
    [sig-windows] Windows volume mounts [Feature:Windows] check volume mount permissions container should have readOnly permissions on hostMapPath

 SUCCESS! 463.260217ms ESC[38;5;228mYou're using deprecated Ginkgo functionality:ESC[0m
ESC[38;5;228m=============================================ESC[0m
  ESC[38;5;11mImprovements to how reporters emit timeline information means that SuppressProgressReporting is no longer necessary and has been deprecated.ESC[0m

ESC[38;5;243mTo silence deprecations that can be silenced set the following environment variable:ESC[0m
  ESC[38;5;243mACK_GINKGO_DEPRECATIONS=2.9.4ESC[0m

--- PASS: TestE2E (0.70s)
PASS
ok      k8s.io/kubernetes/test/e2e      0.769s
```

jiahuif commented on test/e2e/apimachinery/validatingadmissionpolicy.go:
```diff
@@ -36,7 +36,7 @@ import (
 )
 
-var _ = SIGDescribe("ValidatingAdmissionPolicy [Privileged:ClusterAdmin][Alpha]", feature.ValidatingAdmissionPolicy, func() {
+var _ = SIGDescribe("ValidatingAdmissionPolicy [Privileged:ClusterAdmin]", feature.ValidatingAdmissionPolicy, func() {
```
Yes it can be removed. The tag is to filter out all alpha features in job `pull-kubernetes-e2e-kind-alpha-features`. Automatic stability tagging is better than manual for sure. I will remember to update the job once this PR merges.

jiahuif replied:
The feature gate is `feature.ValidatingAdmissionPolicy`. I see you've already added `feature.ValidatingAdmissionPolicy` as the optional second parameter. That's a great change.

pohly replied:
There's a slight difference between `Feature` and `FeatureGate`: only the latter have stability levels and eventually get removed when the feature reaches GA, whereas `Feature` is a permanent tag that identifies some functionality. The pending KEP tries to clarify that.

I am going to use `FeatureGate`, so the test now is:
```
    [sig-api-machinery] ValidatingAdmissionPolicy [Privileged:ClusterAdmin] [FeatureGate:ValidatingAdmissionPolicy] [Alpha] should validate against a Deployment
```

This doesn't break the job.

aojea commented:
I really like this direction, consolidating the tags , but it will be good to have ack from downstream CI consumers since will impact them

/cc @soltysh @deads2k @BenTheElder @spiffxp @dims

aojea commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -67,6 +125,301 @@ func AnnotatedLocationWithOffset(annotation string, offset int) types.CodeLocati
+	return registerInSuite(ginkgo.It, text, args)
+}
+
+// It is a shorthand for the corresponding package function.
```
naive question, why do we want the shorthand?

pohly replied:
It makes lines quite a bit shorter. That's really all there is to it. For the difference, see https://github.com/kubernetes/kubernetes/pull/112894/commits/53c831d6a5708e927934c1a419237abe8d69383b.

I find it much nicer with the shorthand, but can also live with the long form.

aojea replied:
no strong opinion, just that I can see how we may  end with a mix of both forms

aojea commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -67,6 +125,301 @@ func AnnotatedLocationWithOffset(annotation string, offset int) types.CodeLocati
+
+// registerInSuite is the common implementation of all wrapper functions. It
+// expects to be called through one intermediate wrapper.
+func registerInSuite(ginkgoCall func(text string, args ...interface{}) bool, text string, args []interface{}) bool {
```
should we have an unit test to "protect" us of future changes here that may break the format?

pohly replied:
Unit tests for `registerInSuite` certainly can be added. It's complex enough that this is worthwhile. I can add them once we agree on the concept.

pohly replied:
I added unit tests for all new code.

aojea commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -123,6 +124,26 @@ func AnnotatedLocationWithOffset(annotation string, offset int) types.CodeLocati
 	return codeLocation
 }
 
+// SIGDescribe returns a wrapper function for ginkgo.Describe which injects
+// the SIG name as annotation. The parameter should be lowercase with
+// no spaces and no sig- or SIG- prefix.
+func SIGDescribe(sig string) func(string, ...interface{}) bool {
+	if !sigRE.MatchString(sig) || strings.HasPrefix(sig, "sig-") {
+		panic(fmt.Sprintf("SIG label must be lowercase, no spaces and no sig- prefix, got instead: %q", sig))
+	}
```
this is a strong convention and panicing may cause some developer bad UX, because it will be hard to identify the problem ... a linter perhaps?

pohly replied:
You got me here :sweat_smile: Doing the validation inline with panic is trivial, so I went for that first. But you are right, the panic and the root cause for it are not obvious. My plan (not fully worked out yet) was to have an invocation of the test suites as part of `pull-kubernetes-verify` which as clearly as possible shows what problems exist in the suite.

However, writing a linter that handles all cases will not be possible and a lot of work. What it can't handle are calls where the strings are computed and only known at runtime.

pohly replied:
No more panics...

aojea replied:
ok, found the original comment hahah, this is from Jul, my memory has shorter buffer XD

aojea commented on test/e2e/windows/framework.go:
```diff
@@ -17,19 +17,26 @@ limitations under the License.
+	if len(args) == 0 {
+		panic("need additional parameters")
+	}
+	// Callback must come last.
+	body := args[len(args)-1].(func())
+	args[len(args)-1] = func() {
```
is this convention guaranteed? this looks pretty complex to understand, lot of magic here

aojea replied:
why is different though? because of the BeforeEach?

I think that the comment should assume readers don't have any idea of how ginkgo works,

pohly replied:
> Is this convention guaranteed?

It's true for all callers of this particular `SIGDescribe` and also used everywhere else.

> why is different though? because of the BeforeEach?

Yes. That has to be added before the original callback.

pohly replied:
I got dissatisfied with the amount of magic required here. I've added a second commit with a simpler solution (wrap the callback to inject the `BeforeEach`), albeit that now touches more files.

@aojea: I'll let you decide what you prefer.

aojea replied:
I found this last iteration simpler and less invasive

aojea commented on test/e2e/feature/feature.go:
```diff
@@ -0,0 +1,132 @@
+)
+
+func init() {
+	// This prevents adding additional ad-hoc features in tests.
```
didn't have time to go through the whole PR logic, do you mind explain how this prevents that?

pohly replied:
Once this package is done with registering the pre-defined features, it locks the registry. Tests that import this package then cannot add more of them.

There is a loophole: a test that doesn't import the package could define features during its own init and, depending on the order in which packages get initialized, might succeed with that. This is only a problem for init. Describe callbacks get invoked after the global init, so something like this will always fail:
```
_ = Describe("foo", func() {
   It("bar", framework.WithFeature("my-custom-feature"), func() { ...})
})
```

We can also drop this locking and rely instead on code reviews to prevent a proliferation of arbitrary feature tags. I'm open for suggestions here.

aojea replied:
everything that is not automated is doomed to fail XD

aojea commented on test/e2e/feature/feature.go:
```diff
@@ -0,0 +1,132 @@
+*/
+
+// Package feature contains pre-defined features used by test/e2e and/or
+// test/e2e_node.
```
how do we keep this in sync with the `pkg/features`?

didn't we talked about using those values directly so tests and feature gates are in sync by code?

pohly replied:
We did, and [you said](https://github.com/kubernetes/kubernetes/pull/112894#discussion_r996424246):
> I think that we should not create this relation Feature <---> FeatureGate, those are independent things that some times be related, a FeatureGate doesn't have to be related to a Feature necessarily.

This file here is about features, not feature gates (= `pkg/features`), and thus intentionally not forced to be in sync.

aojea replied:
oh, you caught me there, I missed that detail, so we call Feature something like NetworkPolicy or Ingress, sorry about that, I stand on my previous position :)

pohly replied:
I agree, we should have that separation. But nonetheless it is confusing and we'll need very good guidance in the pending KEP and the framework about when to use `FeatureGate` and when `Feature`.

I also still wonder whether we can come up with a different name than `Feature`. It's just too close to `FeatureGate`.

aojea replied:
@SergeyKanzhelev you will be interested on this PR, since you started the discussion here https://github.com/kubernetes/enhancements/pull/3042/files#r988729747 :)

aojea commented:
There are some topics that will impact all sigs and the developer experience and may need some further discussion, it will be good to broadcast these changes to a wider audience, but I'm afraid that before code freeze next week everybody interested will be very busy

I think there are some parts we can get in since are mostly mechanical changes though

pohly commented:
Any suggestion on which parts I should keep to merge it? I can still rework the PR today, but then will be on vacation.

We shouldn't rush it. So perhaps we just wait, clarify during code freeze, and then merge at the start of the next development cycle?

pohly commented:
Looking at some of the opens I prefer to wait. The new API calls could be merged, but those are also the ones where some further work is needed around validation.

aojea commented:
I'm going on vacation next week too until august, let's target 1.29 then

pohly committed:
e2e framework: track and report internal bugs

If something goes wrong during the test registration phase, the only solution
so far was to panic. This is not user-friendly and only allows to report one
problem at a time.

If initialization can continue, then a better solution is to record a bug,
continue, and then report all bugs together.

This also works when just listing tests. The new verify-e2e-suites.sh uses that
to check all test suites (identified as "packages that call
framework.AfterReadingAllFlags", with some exceptions) as part of
pull-kubernetes-verify.

Example output for a fake

    framework.RecordBug(framework.NewBug("fake bug during SIGDescribe", 0))

in test/e2e/storage/volume_metrics.go:
```
$ hack/verify-e2e-suites.sh
go version go1.21.1 linux/amd64
ERROR: E2E test suite invocation failed for test/e2e.
   ERROR: E2E suite initialization was faulty, these errors must be fixed:
   ERROR: test/e2e/storage/volume_metrics.go:49: fake bug during SIGDescribe
E2E suite test/e2e_kubeadm passed.
E2E suite test/e2e_node passed.
```

aojea commented on test/conformance/testdata/conformance.yaml:
```diff
@@ -186,7 +186,7 @@
   codename: '[sig-api-machinery] CustomResourceDefinition resources [Privileged:ClusterAdmin]
     Simple CustomResourceDefinition creating/deleting custom resource definition objects
-    works  [Conformance]'
+    works [Conformance]'
```
this requires special approvers, you better split it in a new PR for faster approval?

pohly replied:
Will do.

pohly replied:
https://github.com/kubernetes/kubernetes/pull/120959

aojea commented on test/e2e/framework/bugs.go:
```diff
@@ -0,0 +1,108 @@
+//
+// This can be used instead of raising a panic. Then all bugs can be reported
+// together instead of failing after the first one.
+func RecordBug(bug Bug) {
```
I personally don't like this behavior, on CIs that runs thousands of jobs per day this can hide important bugs or go unnoticed ... a panic makes it clear something is wrong.

> Then all bugs can be reported  together instead of failing after the first one.

How many of these cases did we have? I prefer to have more evidence that this is a real problem before adding more compelxity to the framework

pohly replied:
> I personally don't like this behavior, on CIs that runs thousands of jobs per day this can hide important bugs or go unnoticed ... a panic makes it clear something is wrong.

You didn't like the panic before ("panicing may cause some developer bad UX"). Also, why should this go unnoticed? It's going to be a test failure in a blocking job (more than one, actually) and each manual invocation of the E2E suite. This will be impossible to miss.

> How many of these cases did we have?

All of the white space issues in the first commit, and more once I start enforcing that inline `[]` must also be tags. You were right, panicking is a bad UX: run E2E suite, fix the one issue which caused the panic, run again, and so on.

Of course this is only during the transition and one could argue that the only developer who is going to suffer is me :sweat_smile: Later, other developers will probably only see a panic while they develop a new test and do not register it correctly. However, I am not sure how visible a panic will be in a Prow job. It might get hidden by the Ginkgo CLI which then only reports "workers failed to start".

pohly replied:
I checked, ginkgo reports the panic. But the UX is definitely worse. This is with `RecordBug` raising a panic instead of the current code and the whitespace commit reverted:
```console
$ _output/bin/ginkgo -p ./test/e2e
Ginkgo timed out waiting for all parallel procs to report back
Test suite: e2e (./test/e2e)

This occurs if a parallel process exits before it reports its results to the
Ginkgo CLI.  The CLI will now print out all the stdout/stderr output it's
collected from the running processes.  However you may not see anything useful
in these logs because the individual test processes usually intercept output to
stdout/stderr in order to capture it in the spec reports.

You may want to try rerunning your test suite with
--output-interceptor-mode=none to see additional output here and debug your
suite.
  
Output from proc 1:
  panic: trailing or leading spaces are unnecessary and need to be removed: "[sig-storage] vcp at scale [Feature:vsphere] "

  goroutine 1 [running]:
  k8s.io/kubernetes/test/e2e/framework.RecordBug({{0x801ddff, 0x4c}, 0x3b, {0xc000bf8900, 0x72}})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/bugs.go:43 +0x54
  k8s.io/kubernetes/test/e2e/framework.registerInSuite(0x6cc4808, {0xc000bd6f30?, 0x2d?}, {0xc0008e4aa0?, 0x2, 0xc000cfa9c0?})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:247 +0xb14
  k8s.io/kubernetes/test/e2e/framework.SIGDescribe.func1({0x6a76de6, 0x1f}, {0xc000cfa9c0, 0x1, 0x1})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:141 +0x246
  k8s.io/kubernetes/test/e2e/storage/vsphere.init()
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/vsphere/vsphere_scale.go:59 +0x1af

Output from proc 2:
  panic: trailing or leading spaces are unnecessary and need to be removed: "[sig-storage] vcp at scale [Feature:vsphere] "
...
Output from proc 35:
  panic: trailing or leading spaces are unnecessary and need to be removed: "[sig-storage] vcp at scale [Feature:vsphere] "

  goroutine 1 [running]:
  k8s.io/kubernetes/test/e2e/framework.RecordBug({{0x801ddff, 0x4c}, 0x3b, {0xc0001abe00, 0x72}})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/bugs.go:43 +0x54
  k8s.io/kubernetes/test/e2e/framework.registerInSuite(0x6cc4808, {0xc0006d05d0?, 0x2d?}, {0xc00055b340?, 0x2, 0xc000a919d0?})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:247 +0xb14
  k8s.io/kubernetes/test/e2e/framework.SIGDescribe.func1({0x6a76de6, 0x1f}, {0xc000a919d0, 0x1, 0x1})
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/ginkgowrapper.go:141 +0x246
  k8s.io/kubernetes/test/e2e/storage/vsphere.init()
  	/nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/vsphere/vsphere_scale.go:59 +0x1af

** End **
Ginkgo ran 1 suite in 22.389893721s

Test Suite Failed
```

aojea replied:
>  panic: trailing or leading spaces are unnecessary and need to be removed

I can't remember the exact context, but I mean, panic is used when someting went unexpectedly wrong and we are using it here to warn about "linting" on title, sounds like a big hammer

> Of course this is only during the transition and one could argue that the only developer who is going to suffer is me 😅

can't we make this more subtle? if is only for the transition ...

pohly replied:
A developer may always do something wrong also after the transition. If we make error reporting less obvious (for example, an error log message), then there is a considerable risk of regressions that will go unnoticed.

Is that what you meant with "subtle"?

I think the E2E suite should fail to run if test registration is not correct. I prefer the current approach with `RecordBug` because (as you said yourself :grin:) `panic` is not user-friendly. But if you find the current approach too complex, then I'm also fine with `panic`.

pohly replied:
With the new `ginkgo.PreviewSpecs`, bugs can be detected before even allowing the test suite to start. I think that's nicer and prevents wasting resources on testing a flawed PR, so I've switched to that.

aojea replied:
sgtm

aojea commented:
I will start small with the tags and lists and avoid the other more complex changes

pohly commented:
It looks like https://github.com/onsi/ginkgo/issues/1225 is going to be solved soon. Let's wait for it and use it. However, we can continue to discuss other aspects, like panic vs. `ReportBug`.

/hold

pohly committed:
e2e storage: remove useless info message

Always printing "Enabling in-tree volume drivers" whenever the E2E suite is
initializing doesn't provide any useful information and makes output of the
upcoming -list-tests look weird.

aojea commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -67,6 +125,322 @@ func AnnotatedLocationWithOffset(annotation string, offset int) types.CodeLocati
+
+	// Now that we have the final offset, we can record bugs.
+	if haveEmptyStrings {
+		RecordBug(NewBug("empty strings as separators are unnecessary and need to be removed", int(offset)))
```
what's the deal with the spaces and empty strings, I can't remember

pohly replied:
The final `strings.Join(" ", components)` will add two spaces around any empty string - just a nit, but not nice.

aojea commented:
you have some error with the vendor folder

> +
> +func withCancelCause(parent context.Context) (context.Context, func(error)) {
> +	return context.WithCancelCause(parent)

This LGTM from a quick look, but others should check

/assign @SergeyKanzhelev @BenTheElder

pohly commented:
> you have some error with the vendor folder

Let's merge the gingko+gomega update separately in https://github.com/kubernetes/kubernetes/pull/121082, then I can remove that part from this PR.

akhilerm commented:
Retriggering the below presubmits as there has been a GCE image change in the jobs in https://github.com/kubernetes/test-infra/pull/31008
```
pull-kubernetes-e2e-gci-gce-ingress
pull-kubernetes-e2e-ubuntu-gce-network-policies
pull-kubernetes-e2e-gci-gce-ipvs
```

akhilerm commented:
/test pull-kubernetes-e2e-gci-gce-ingress
/test pull-kubernetes-e2e-ubuntu-gce-network-policies
/test pull-kubernetes-e2e-gci-gce-ipvs

aojea commented:
/lgtm
/approve

Thanks @pohly for the multiple iterations ,

aojea commented:
@pohly should we announce this to the wide community for awareness?
